### PR TITLE
fix implementation of IBFEInstrumentPanel

### PR DESCRIPTION
**Pull request by boyceg**
### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

**Comment by drwells** [comment]
I'm all for doing this, but right now I'm trying to write some tests for `IBFEInstrumentPanel` to verify that things work. Let me see if I can finish that today and then we can start refactoring.

**Comment by boyceg** [comment]
Unfortunately `IBFEInstrumentPanel` is totally broken, so tests will not help.

**Comment by drwells** [comment]
Does this PR actually fix it or just get rid of the custom kernels etc.?

**Comment by boyceg** [comment]
This PR is WIP. It will eventually fix it.

**Comment by drwells** [comment]
I tidied-up my test code and its available at https://github.com/drwells/IBAMR/tree/instrument-panel-test (see https://github.com/drwells/IBAMR/blob/instrument-panel-test/tests/IBFE/instrument_panel_01.cpp).

Interestingly, I get pretty good results for the meters with my tests. Here's a comparison of the meter code flux vs. the one I compute with surface quadrature on the original mesh:
```
[drwells@archway IBFE]$ cat viz_inst3d_meters/flux.dat                      
1 2.1796701674124 2.1796701674124
1 0.493775870234542 0.493775870234542

[drwells@archway IBFE]$ cat output 
Number of elements: 64

[...]

bottom flux = -2.18514
top flux = 2.18514
bottom flux = -0.49107
top flux = 0.49107
```

**Comment by boyceg** [comment]
@drwells the bugs are related to mesh motions that result in any rotations or dilations of meter mesh elements.

**Comment by drwells** [comment]
I can set up tests in those cases pretty easily - let me know. Per the discussion on slack I'm going to go back to the other fun set of bugs (outernodal data).

**Comment by boyceg** [comment]
> I can set up tests in those cases pretty easily - let me know. Per the discussion on slack I'm going to go back to the other fun set of bugs (outernodal data).

If you have a simple flow that causes the meter to "expand" or "contract", that would be a good one. The other thing that comes to mind is something that generates a rigid-body motion.

**Comment by drwells** [comment]
For the purposes of these tests I'm just manually moving the test so we can do whatever we want.

**Comment by cpuelz** [comment]
I've added code to improve the orientation of the nodes for the meter meshes in fix-meter-orientation branch. For each nodeset I build a map from a node ID to the set of elements that contain the node with that ID. I then use this map to determine if two nodes are adjacent by intersecting their set of elements.

**Comment by drwells** [comment]
I updated the test code - I also switched to using a Taylor-Green vortex sheet. The meter code now disagrees with the custom flux code in the rigid translation case.

**Comment by boyceg** [comment]
> I've added code to improve the orientation of the nodes for the meter meshes in the fix-meter-orientation branch. For each nodeset I build a map from a node ID to the set of elements that contain the node with that ID. I then use this map to determine if two nodes are adjacent by intersecting their set of elements.

Should I merge that into my branch?

**Comment by cpuelz** [comment]
> > I've added code to improve the orientation of the nodes for the meter meshes in the fix-meter-orientation branch. For each nodeset I build a map from a node ID to the set of elements that contain the node with that ID. I then use this map to determine if two nodes are adjacent by intersecting their set of elements.
> 
> Should I merge that into my branch?

yes, go for it. it is only a change to ```IBFEInstrumentPanel::initializeHierarchyIndependentData```.

File: src/IB/IBFEInstrumentPanel.cpp [hunk 1]
```diff
@@ -927,161 +459,314 @@ IBFEInstrumentPanel::getNumberOfMeterMeshes() const
...
+        Node* node = &d_meter_meshes[meter_idx]->node_ref(i);
+        for (int d = 0; d < NDIM; ++d)
         {
-            const int vel_dof_idx = node->dof_number(velocity_sys_num, d, 0);
-            velocity_solution.set(vel_dof_idx, velocity_coords(vel_dof_idx));
-            const int disp_dof_idx = node->dof_number(displacement_sys_num, d, 0);
-            displacement_solution.set(disp_dof_idx, displacement_coords(disp_dof_idx));
+            (*node)(d) = x_meter_dofs[NDIM * i + d];
+            const int u_dof_idx = d_u_dof_idx[meter_idx][i][d];
+            u_meter_vec->set(u_dof_idx, u_meter_dofs[NDIM * i + d]);
```
Comment (cpuelz) [inline src/IB/IBFEInstrumentPanel.cpp:561]: maybe here you are setting the meter velocity DOF with the mesh velocity DOF?

Comment (boyceg) [reply src/IB/IBFEInstrumentPanel.cpp:562]: Just pushed a fix.

**Comment by boyceg** [comment]
I think this works.

File: tests/IBFE/instrument_panel_01_3d.output [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+IBHierarchyIntegrator::initializePatchHierarchy(): tag_buffer = 3
+INSStaggeredHierarchyIntegrator::initializeCompositeHierarchyData():
+  projecting the interpolated velocity field
+INSStaggeredHierarchyIntegrator::regridProjection(): regrid projection solve residual norm        = 2.58993e-05
+
+
+Writing visualization files...
+
+test 0
+Warning in file ``/Users/boyceg/code/IBAMR/lib/../src/IB/IBFEInstrumentPanel.cpp'' at line 658
```
Comment (drwells) [inline tests/IBFE/instrument_panel_01_3d.output:14]: Mind if I push a change to suppress this warning? This path isn't going to work on other people's machines :)

Comment (boyceg) [reply tests/IBFE/instrument_panel_01_3d.output:14]: Please do

Comment (boyceg) [reply tests/IBFE/instrument_panel_01_3d.output:14]: Whoops --- it doesn't even work on my own machine!

Comment (drwells) [reply tests/IBFE/instrument_panel_01_3d.output:14]: I'm working on it now - we have a class `TestAppender` in `tests/tests.h` which works around exactly this issue with SAMRAI.

Comment (drwells) [reply tests/IBFE/instrument_panel_01_3d.output]: I fixed the warnings and a few other minor things.

File: tests/IBFE/instrument_panel_01.cpp [hunk 3]
```diff
@@ -515,23 +518,23 @@ main(int argc, char** argv)
             const double bottom_meter_flux = instrument_panel.getMeterFlowRates()[0];
             const double top_meter_flux = instrument_panel.getMeterFlowRates()[1];
 
-            plog << "bottom meter flux = " << bottom_analytic_flux << std::endl;
-            plog << "top meter flux = " << top_analytic_flux << std::endl;
-
-            plog << "bottom flux absolute difference = " << abs((bottom_analytic_flux - bottom_meter_flux)) << std::endl;
-            plog << "top flux absolute difference = " << abs((top_analytic_flux - top_meter_flux)) << std::endl;
+            plog << "bottom meter flux = " << bottom_meter_flux << std::endl;
+            plog << "top meter flux = " << top_meter_flux << std::endl;
```
Comment (drwells) [inline tests/IBFE/instrument_panel_01.cpp:522]: :)

File: tests/IBFE/instrument_panel_01.cpp [hunk 4]
```diff
@@ -463,20 +467,19 @@ main(int argc, char** argv)
...
-                    double z;
 #if LIBMESH_VERSION_LESS_THAN(1, 7, 0)
-                    z = side_ptr->centroid()(2);
+                    const double z = side_ptr->centroid()(2);
 #else
-                    z = side_ptr->vertex_average()(2);
+                    const double z = side_ptr->vertex_average()(2);
 #endif
-                    if (abs(z - (data_time + 0.0)) < 1.0e-5)
+                    if (std::abs(z - (data_time + 0.0)) < 1.0e-5)
```
Comment (drwells) [inline tests/IBFE/instrument_panel_01.cpp:475]: `abs()` is a C function defined for integers only - I prefer to always write `std::abs()` to make sure we are talking about the overloaded version in C++.

**Review by drwells — APPROVED** [review]
Looks good - we should write some more comprehensive tests but we can handle that in a followup PR. I added a changelog entry.

Would you mind squashing these commits together?

**Comment by boyceg** [comment]
> Looks good - we should write some more comprehensive tests but we can handle that in a followup PR. I added a changelog entry.
> 
> Would you mind squashing these commits together?

Done, I think!

**Comment by boyceg** [comment]
> > Looks good - we should write some more comprehensive tests but we can handle that in a followup PR. I added a changelog entry.
> > Would you mind squashing these commits together?
> 
> Done, I think!

Looks like @cpuelz is getting credit for this PR in the GitHub bean counting. 😎

**Comment by boyceg** [comment]
@drwells it looks like I missed one commit in the squash.

**Comment by boyceg** [comment]
Fixed now, although I guess it is too late for `master.

**Comment by drwells** [comment]
meh, 2 is better than 30 if we want to read the history later. I'm not worried about it.

**Comment by cpuelz** [comment]
> > > Looks good - we should write some more comprehensive tests but we can handle that in a followup PR. I added a changelog entry.
> > > Would you mind squashing these commits together?
> > 
> > 
> > Done, I think!
> 
> Looks like @cpuelz is getting credit for this PR in the GitHub bean counting. sunglasses

I should be docked some beans for the initial implementation :-)

**Comment by boyceg** [comment]
> > > > Looks good - we should write some more comprehensive tests but we can handle that in a followup PR. I added a changelog entry.
> > > > Would you mind squashing these commits together?
> > > 
> > > 
> > > Done, I think!
> > 
> > 
> > Looks like @cpuelz is getting credit for this PR in the GitHub bean counting. sunglasses
> 
> I should be docked some beans for the initial implementation :-)

I was looking into how to do that but then David merged. 😎

**Pull request merged**